In [132]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import warnings
from bokeh.io import output_file, show

In [133]:
warnings.filterwarnings('ignore')

In [134]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [135]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [136]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [137]:
df_market_data.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                           1.08388                     7.60278   
ethereum                          0.22392                    10.38134   
tether                           -0.21173                     0.04935   
ripple                           -0.37819                    -0.60926   
bitcoin-cash                      2.90585                    17.09717   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                           6.57509                      7.67258   
ethereum                          4.80849                      0.13169   
tether                            0.00640                     -0.04237   
ripple                            2.24984                      0.23455   
bitcoin-cash                     14.75334                     15.74903   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                          -3.25185                      83.51840   
ethereum                        -12.88890                     186.77418   
tether                            0.28037                      -0.00542   
ripple                          -17.55245                      39.53888   
bitcoin-cash                    -13.71793                      21.66042   

              price_change_percentage_1y  
coin_id                                   
bitcoin                         37.51761  
ethereum                       101.96023  
tether                           0.01954  
ripple                         -16.60193  
bitcoin-cash                    14.49384

In [138]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()
scaled_data = scaler.fit_transform(df_market_data)


In [139]:
# Create a DataFrame with the scaled data
scaled_market_data = pd.DataFrame(scaled_data, columns=df_market_data.columns, index=df_market_data.index)
# Copy the crypto names from the original DataFrame
crypto_names = df_market_data.index.copy()

# Set the coin_id column as index
scaled_market_data.index = crypto_names

# Display the scaled DataFrame
print(crypto_names)

Index(['bitcoin', 'ethereum', 'tether', 'ripple', 'bitcoin-cash',
       'binancecoin', 'chainlink', 'cardano', 'litecoin', 'bitcoin-cash-sv',
       'crypto-com-chain', 'usd-coin', 'eos', 'monero', 'tron', 'tezos', 'okb',
       'stellar', 'cosmos', 'cdai', 'neo', 'wrapped-bitcoin', 'leo-token',
       'huobi-token', 'nem', 'binance-usd', 'iota', 'vechain', 'zcash',
       'theta-token', 'dash', 'ethereum-classic', 'ethlend', 'maker', 'havven',
       'omisego', 'celsius-degree-token', 'ontology', 'ftx-token', 'true-usd',
       'digibyte'],
      dtype='object', name='coin_id')


---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [140]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12))
print(k_values)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]


In [141]:
scaled_market_data

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [142]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop: 
for k in k_values:
# 1. Create a KMeans model using the loop counter for the n_clusters
    kmeans_model = KMeans(n_clusters=k, random_state=0)
# 2. Fit the model to the data using `df_market_data_scaled`
    kmeans_model.fit(df_market_data)
# 3. Append the model.inertia_ to the inertia list
    inertia.append(kmeans_model.inertia_)

In [143]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = pd.DataFrame({
    'k_values': k_values,
    'inertia': inertia
})

# Create a DataFrame with the data to plot the Elbow curve


In [144]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_plot_1 = elbow_data.hvplot.line(
    x='k_values',
    y='inertia',
    title='Elbow Curve',
    xlabel='Number of Clusters (k)',
    ylabel='Inertia',
    line_color='blue',
)

hvplot.show(elbow_plot_1)

Launching server at http://localhost:56670


#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:**  6 Clusters would be the best value for 'k'.

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [145]:
scaled_market_data_numeric = scaled_market_data.select_dtypes(include=['float64', 'int64'])


In [146]:
# Initialize the K-Means model using the best value for k
best_k = 6
kmeans_model = KMeans(n_clusters=best_k, random_state=0)



In [147]:
# Fit the K-Means model using the scaled DataFrame
kmeans_model.fit(scaled_market_data_numeric)


KMeans(n_clusters=6, random_state=0)

In [148]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
cluster_predictions = kmeans_model.predict(scaled_market_data_numeric)

scaled_market_data_numeric['cluster'] = cluster_predictions

# Print the resulting array of cluster values.
print(cluster_predictions)

[2 2 0 0 5 2 5 2 2 0 0 0 0 5 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 4 2 0 1 2 0 0 3
 0 2 0 0]


In [149]:
# Create a copy of the scaled DataFrame
copy_scaled = scaled_market_data_numeric.copy()

In [150]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
copy_scaled['cluster'] = cluster_predictions

# Display the copy of the scaled DataFrame
copy_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  cluster  
coin_id                                            
bitcoin                        -0.251637        2  
ethereum                       -0.199352        2  
tether                         -0.282061        0  
ripple                         -0.295546        0  
bitcoin-cash                   -0.270317        5

In [152]:
# Create a scatter plot using hvPlot by setting
scatter_plot = copy_scaled.hvplot.scatter(    
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
    x='price_change_percentage_24h',
    y='price_change_percentage_7d',
# Color the graph points with the labels found using K-Means and
    c='cluster',
# add the crypto name in the `hover_cols` parameter to identify
    hover_cols=['coin_id'],
# the cryptocurrency represented by each data point.
    title='K-Means Clustering of Cryptocurrencies',
    xlabel='Price Change Percentage (24h)',
    ylabel='Price Change Percentage (7d)',
    size=100, 
    cmap='viridis'
)
hvplot.show(scatter_plot)

Launching server at http://localhost:56689


---

### Optimize Clusters with Principal Component Analysis.

In [153]:
# Create a PCA model instance and set `n_components=3`.
pca_model = PCA(n_components=3)

In [154]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
principal_components = pca_model.fit_transform(scaled_market_data_numeric)
principal_df = pd.DataFrame(principal_components, columns=[f'PC{i+1}' for i in range(3)])
principal_df['coin_id'] = scaled_market_data_numeric.index
# View the scaled PCA data
principal_df.head()

PC1       PC2       PC3       coin_id
0  1.148906 -0.806582 -0.150162       bitcoin
1  1.058585 -0.591759 -0.718260      ethereum
2 -1.158291 -0.339164  0.681214        tether
3 -1.159885 -0.367816  0.531758        ripple
4  4.290020 -1.729224 -1.455448  bitcoin-cash

In [155]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
explainded_variance = pca_model.explained_variance_ratio_

variance_df = pd.DataFrame({
    'Principal Component': [f'PC{i+1}' for i in range(len(explainded_variance))],
    'Explained Variance Ratio': explainded_variance
})

variance_df.head()

Principal Component  Explained Variance Ratio
0                 PC1                  0.418720
1                 PC2                  0.278413
2                 PC3                  0.182964

In [156]:
total_explained_variance = sum(explainded_variance)
print(f'Total Explained Variance of the Three Principal Components: {total_explained_variance:.4f}')

Total Explained Variance of the Three Principal Components: 0.8801


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** Total Explained Variance of the Three Principal Components: 0.8801

In [157]:
# Create a new DataFrame with the PCA data.
pca_df = pd.DataFrame(principal_components, columns=[f'PC{i+1}' for i in range(3)])

# Copy the crypto names from the original scaled DataFrame
crypto_names = scaled_market_data_numeric.index.tolist()
pca_df['coin_id'] = crypto_names

# Set the coin_id column as index
pca_df.set_index('coin_id', inplace=True)

# Display the scaled PCA DataFrame
pca_df.head()

PC1       PC2       PC3
coin_id                                   
bitcoin       1.148906 -0.806582 -0.150162
ethereum      1.058585 -0.591759 -0.718260
tether       -1.158291 -0.339164  0.681214
ripple       -1.159885 -0.367816  0.531758
bitcoin-cash  4.290020 -1.729224 -1.455448

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [158]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1,12))
print(k_values)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]


In [159]:
# Create an empty list to store the inertia values
inertia_values = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
for k in k_values:
# 1. Create a KMeans model using the loop counter for the n_clusters
    kmeans_model = KMeans(n_clusters=k, random_state=0)
# 2. Fit the model to the data using `df_market_data_pca`
    kmeans_model.fit(pca_df)
# 3. Append the model.inertia_ to the inertia list
    inertia_values.append(kmeans_model.inertia_)

In [160]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    'k_values': k_values,
    'inertia_values': inertia_values
}
# Create a DataFrame with the data to plot the Elbow curve
elbow_df = pd.DataFrame(elbow_data)

In [162]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_plot_2 = elbow_df.hvplot.line(
    x='k_values',
    y='inertia_values',
    title='Elbow Curve',
    xlabel='Number of Clusters (k)',
    ylabel='Inertia',
    line_color='blue',
)

hvplot.show(elbow_plot_2)

Launching server at http://localhost:56707


#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for 'k' would be 7 clusters.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** It does differ but by only 1 cluster.

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [163]:
# Initialize the K-Means model using the best value for k
best_k_value = 7
kmeans_model = KMeans(n_clusters=best_k, random_state=0)

In [164]:
# Fit the K-Means model using the PCA data
kmeans_model.fit(pca_df)

KMeans(n_clusters=6, random_state=0)

In [165]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
cluster_predictions_pca = kmeans_model.predict(pca_df)
pca_df['cluster'] = cluster_predictions_pca

# Print the resulting array of cluster values.
print(pca_df)

                           PC1       PC2       PC3  cluster
coin_id                                                    
bitcoin               1.148906 -0.806582 -0.150162        1
ethereum              1.058585 -0.591759 -0.718260        1
tether               -1.158291 -0.339164  0.681214        0
ripple               -1.159885 -0.367816  0.531758        0
bitcoin-cash          4.290020 -1.729224 -1.455448        4
binancecoin           1.593077 -0.853364 -0.030404        1
chainlink             3.707123 -0.719496 -3.020461        4
cardano               1.387119 -0.553302 -0.902328        1
litecoin              0.917647 -0.765043 -0.512190        1
bitcoin-cash-sv      -1.131914 -0.657312  0.333537        0
crypto-com-chain     -2.107618  0.128015  0.520788        0
usd-coin             -1.169535 -0.342306  0.691306        0
eos                  -1.373547 -0.526198  0.385925        0
monero                4.975308 -0.730971 -0.535519        4
tron                 -0.957828 -0.326046

In [166]:
# Create a copy of the scaled PCA DataFrame
copied_pca = pca_df.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
copied_pca['predicted clusters'] = cluster_predictions_pca
# Display the copy of the scaled PCA DataFrame
copied_pca.head()

PC1       PC2       PC3  cluster  predicted clusters
coin_id                                                                
bitcoin       1.148906 -0.806582 -0.150162        1                   1
ethereum      1.058585 -0.591759 -0.718260        1                   1
tether       -1.158291 -0.339164  0.681214        0                   0
ripple       -1.159885 -0.367816  0.531758        0                   0
bitcoin-cash  4.290020 -1.729224 -1.455448        4                   4

In [168]:
# Create a scatter plot using hvPlot by setting
scatter_plot_pca = pca_df.hvplot.scatter(
# `x="PC1"` and `y="PC2"`.
    x='PC1',
    y='PC2',
# Color the graph points with the labels found using K-Means and
    c='cluster',
# add the crypto name in the `hover_cols` parameter to identify
    hover_cols=['cluster'],
# the cryptocurrency represented by each data point.
    title='K-Means Clustering of Cryptocurrencies',
    xlabel='Principal Component 1 (PC1)',
    ylabel='Principal Component 2 (PC2)',
    size=100,  
    cmap='viridis'
)
hvplot.show(scatter_plot_pca)

Launching server at http://localhost:56725


### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [170]:
# Composite plot to contrast the Elbow curves
elbow_data_1 = {'k_values': list(range(1, 12)), 'inertia_values': inertia}  # Original
elbow_data_2 = {'k_values': list(range(1, 12)), 'inertia_values': inertia_values}  # PCA

elbow_df_1 = pd.DataFrame(elbow_data_1)
elbow_df_2 = pd.DataFrame(elbow_data_2)

line_plot_1 = elbow_df_1.hvplot.line(
    x='k_values',
    y='inertia_values',
    label='Original Data',
    line_color='blue'
)

line_plot_2 = elbow_df_2.hvplot.line(
    x='k_values',
    y='inertia_values',
    label='PCA Data',
    line_color='orange'
)

composite_plot = line_plot_1 * line_plot_2

composite_plot = composite_plot.opts(
    title='Elbow Curves Comparison',
    xlabel='Number of Clusters (k)',
    ylabel='Inertia',
    legend_position='top_left'
)

hvplot.show(composite_plot)


Launching server at http://localhost:56743


In [172]:
# Composite plot to contrast the clusters
pca_scatter = pca_df.hvplot.scatter(
    x='PC1',
    y='PC2',
    c='cluster',  
    hover_cols=['coin_id', 'cluster'],
    title='K-Means Clustering: PCA Data',
    size=100,
    cmap='viridis'
)

original_scatter = df_market_data.hvplot.scatter(
    x='price_change_percentage_24h',
    y='price_change_percentage_7d',
    c='cluster',  
    hover_cols=['coin_id', 'cluster'],
    title='K-Means Clustering: Original Data',
    size=100,
    cmap='plasma'
)

composite_clusters_plot = pca_scatter * original_scatter

composite_clusters_plot.opts(
    title='Comparison of K-Means Clusters',
    xlabel='Principal Component 1 (PC1) / Price Change 24h',
    ylabel='Principal Component 2 (PC2) / Price Change 7d',
    legend_position='top_left'
)

hvplot.show(composite_clusters_plot)


Launching server at http://localhost:56763


ERROR:bokeh.core.validation.check:E-1001 (BAD_COLUMN_NAME): Glyph refers to nonexistent column name. This could either be due to a misspelling or typo, or due to an expected column being missing. : fill_color='cluster' [no close matches], hatch_color='cluster' [no close matches], line_color='cluster' [no close matches] {renderer: GlyphRenderer(id='p4233', ...)}


#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** It makes the data clearer and make it easier to read the data.